pkg/batch/jobs/notification/domain_event_informer.go

```diff
@@ -0,0 +1,294 @@
+	"go.uber.org/zap"
+	gcodes "google.golang.org/grpc/codes"
+	gstatus "google.golang.org/grpc/status"
+
+	"github.com/bucketeer-io/bucketeer/pkg/batch/jobs"
+	environmentclient "github.com/bucketeer-io/bucketeer/pkg/environment/client"
+	"github.com/bucketeer-io/bucketeer/pkg/errgroup"
+	"github.com/bucketeer-io/bucketeer/pkg/metrics"
+	"github.com/bucketeer-io/bucketeer/pkg/notification/sender"
+	"github.com/bucketeer-io/bucketeer/pkg/pubsub/puller"
+	"github.com/bucketeer-io/bucketeer/pkg/pubsub/puller/codes"
+	"github.com/bucketeer-io/bucketeer/pkg/uuid"
+	environmentproto "github.com/bucketeer-io/bucketeer/proto/environment"
+	domaineventproto "github.com/bucketeer-io/bucketeer/proto/event/domain"
+	notificationproto "github.com/bucketeer-io/bucketeer/proto/notification"
+	senderproto "github.com/bucketeer-io/bucketeer/proto/notification/sender"
+)
+
+var (
+	ErrUnknownSourceType = errors.New("batch-server: domain-event-informer unknown source type")
+)
+
+type options struct {
+	maxMPS                  int
+	numWorkers              int
+	runningDurationPerBatch time.Duration
+	metrics                 metrics.Registerer
+	logger                  *zap.Logger
+}
+
+var defaultOptions = options{
+	maxMPS:     10,
+	numWorkers: 1,
+	logger:     zap.NewNop(),
+}
+
+type Option func(*options)
+
+func WithRunningDurationPerBatch(d time.Duration) Option {
+	return func(opts *options) {
+		opts.runningDurationPerBatch = d
+	}
+}
+
+func WithMaxMPS(mps int) Option {
+	return func(opts *options) {
+		opts.maxMPS = mps
+	}
+}
+
+func WithNumWorkers(n int) Option {
+	return func(opts *options) {
+		opts.numWorkers = n
+	}
+}
+
+func WithMetrics(r metrics.Registerer) Option {
+	return func(opts *options) {
+		opts.metrics = r
+	}
+}
+
+func WithLogger(logger *zap.Logger) Option {
+	return func(opts *options) {
+		opts.logger = logger
+	}
+}
+
+type domainEventInformer struct {
+	environmentClient environmentclient.Client
+	puller            puller.RateLimitedPuller
+	sender            sender.Sender
+	group             errgroup.Group
+	opts              *options
+	logger            *zap.Logger
+}
+
+func NewDomainEventInformer(
+	environmentClient environmentclient.Client,
+	p puller.Puller,
+	sender sender.Sender,
+	opts ...Option) jobs.Job {
+	options := defaultOptions
+	for _, opt := range opts {
+		opt(&options)
+	}
+	if options.metrics != nil {
+		registerMetrics(options.metrics)
+	}
+	return &domainEventInformer{
+		environmentClient: environmentClient,
+		puller:            puller.NewRateLimitedPuller(p, options.maxMPS),
+		sender:            sender,
+		opts:              &options,
+		logger:            options.logger.Named("sender"),
+	}
+}
+
+func (i *domainEventInformer) Run(ctx context.Context) error {
+	i.logger.Info("DomainEventInformer start running")
+	runningDurationCtx, cancel := context.WithTimeout(ctx, i.opts.runningDurationPerBatch)
+	defer cancel()
+	i.group.Go(func() error {
+		return i.puller.Run(runningDurationCtx)
```

Comment:
@cre8ivejp About if we can remove goroutine in `domainEventInformer`, I found that the documentation of [`pubsub Receive` method](https://pkg.go.dev/cloud.google.com/go/pubsub@v1.31.0#Subscription.Receive):

>Receive calls f with the outstanding messages from the subscription. It **_blocks until ctx is done_**, or the service returns a non-retryable error.

So we can't implement this without goroutine.